File: programmers/level2/스킬트리.py
"""
문제 링크: https://programmers.co.kr/learn/courses/30/lessons/49993
문제 접근 방법: 단순 구현, skill에 있는 문자열을 제외한 모든 문자열을 제거하고 skill의 순서대로 skill_tree들이 잘 구성됐는지 하나하나 검사했다. 간단한 구현이긴한데, 코드를 간결하게 작성못한 부분이 조금 아쉬웠다.
"""

def solution(skill, skill_trees):
    alphabet_list = []
    for i in range(65, 91):
        alphabet_list.append(skill.count(chr(i)))
    
    for i in range(len(skill_trees)):
        new_skill = ""
        for j in skill_trees[i]:
            if alphabet_list[ord(j)-65] == 1:
                new_skill += j
        skill_trees[i] = new_skill
    
    answer = 0
    
    for i in range(len(skill_trees)):
        if len(skill_trees[i]) == 0:
            answer+=1
        else:
            k = 0
            answer+=1
            for j in skill_trees[i]:
                if j == skill[k]:
                    k+=1
                else:
                    answer-=1
                    break
        
    return answer

File: programmers/level2/스킬트리.java
import java.util.*;
class Solution {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        
        ArrayList<Integer> alphabetList = new ArrayList<>();
        
        for(int i=65; i<=90; i++) {
            char charI = (char)i;
            int cnt = 0 ;
            for(int j=0; j<skill.length(); j ++ ) {
                if(skill.charAt(j) == charI) 
                    cnt ++;
            }
            alphabetList.add(cnt);
        }
        
        for(int i=0; i< skill_trees.length; i++) {
            String newSkill = "";
            
            for(int j=0; j< skill_trees[i].length(); j++) {
                if(alphabetList.get(skill_trees[i].charAt(j) - 'A') == 1) {
                    newSkill += skill_trees[i].charAt(j);
                }
            }
            skill_trees[i] = newSkill;
        }
        
        
        for(String skillTree : skill_trees) {
            if(skillTree.length() == 0 ) {
                answer ++;
            } else {
                int k =0;
                answer ++;
                for(int j=0; j <skillTree.length();j++) {
                    if(skillTree.charAt(j) == skill.charAt(k) ) k ++;
                    else {
                        answer --;
                        break;
                    }
                }
            }
        }
        
        return answer;
    }
}